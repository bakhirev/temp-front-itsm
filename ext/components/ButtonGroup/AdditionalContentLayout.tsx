import styled from 'styled-components';

const MARGIN_HORIZONTAL = 6;

interface IAdditionalContentLayoutProps {
  left?: boolean;
  right?: boolean;
}

export const AdditionalContentLayout = styled.div<IAdditionalContentLayoutProps>`
  align-items: center;
  display: flex;
  justify-content: center;
  margin-left: ${({ right }: IAdditionalContentLayoutProps) => (right ? MARGIN_HORIZONTAL : 0)}px;
  margin-right: ${({ left }: IAdditionalContentLayoutProps) => (left ? MARGIN_HORIZONTAL : 0)}px;
`;
