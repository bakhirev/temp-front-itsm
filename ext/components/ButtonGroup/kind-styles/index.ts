import { solid } from './solid';
import { outlined } from './outlined';

export const kindStyles = {
  solid,
  outlined,
};
