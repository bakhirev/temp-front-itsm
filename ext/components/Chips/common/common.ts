export type Size = 'big' | 'small';
