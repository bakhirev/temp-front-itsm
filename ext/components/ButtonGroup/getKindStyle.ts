import type { ITheme } from '../themes';

import { kindStyles } from './kind-styles';
import type { Kind } from './Button';

export const getKindStyle = ({
  active,
  disabled,
  kind,
  theme,
}: {
  active: boolean;
  disabled: boolean;
  kind: Kind;
  theme: ITheme;
}) => kindStyles[kind]({ active, disabled, theme });
