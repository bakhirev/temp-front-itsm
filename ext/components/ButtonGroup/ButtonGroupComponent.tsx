import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../common';

import type { Kind } from './Button';

export type Size = 'medium' | 'small' | 'micro';

export const SIZE = {
  medium: 48,
  small: 40,
  micro: 32,
};

interface IButtonGroupComponentProps {
  kind: Kind;
  size: Size;
}

export const ButtonGroupComponent = styled.div<IButtonGroupComponentProps>`
  ${({ kind, size, theme }) => css`
    background: ${kind === 'solid' ? theme.color.neutral[10] : 'transparent'};
    border-radius: ${theme.borderRadius};
    display: flex;
    height: ${SIZE[size]}px;
  `};
`;

ButtonGroupComponent.defaultProps = {
  theme: DEFAULT_THEME,
};
