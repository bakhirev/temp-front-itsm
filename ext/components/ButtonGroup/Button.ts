import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../common';

import { getKindStyle } from './getKindStyle';
import type { Size } from './ButtonGroupComponent';

const PADDING_HORIZONTAL = {
  medium: 24,
  small: 20,
  micro: 16,
};

const BORDER_WIDTH = 1;

export type Kind = 'solid' | 'outlined';

interface IButtonProps {
  active: boolean;
  disabled: boolean;
  kind: Kind;
  size: Size;
}

export const Button = styled.div<IButtonProps>`
  ${({ active, disabled, kind, size, theme }) => css`
    align-items: center;
    border-radius: ${theme.borderRadius};
    border-style: solid;
    border-width: ${BORDER_WIDTH}px;
    box-sizing: border-box;
    color: ${disabled ? theme.color.neutral[30] : theme.color.neutral[50]};
    cursor: ${disabled ? 'default' : 'pointer'};
    display: flex;
    padding: 0 ${PADDING_HORIZONTAL[size] - BORDER_WIDTH}px;
    position: relative;

    & div svg {
      fill: ${disabled ? theme.color.neutral[30] : theme.color.neutral[50]};
    }

    &:focus {
      outline: none;

      &::after {
        border-radius: ${theme.borderRadius};
        border-style: solid;
        border-width: 1px;
        bottom: 2px;
        content: '';
        display: block;
        left: 2px;
        position: absolute;
        right: 2px;
        top: 2px;
      }
    }

    ${getKindStyle({ active, disabled, kind, theme })};
  `};
`;

Button.defaultProps = {
  theme: DEFAULT_THEME,
};
