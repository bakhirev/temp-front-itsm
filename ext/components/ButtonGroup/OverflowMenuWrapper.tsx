import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../common';

import { SIZE } from './ButtonGroupComponent';
import type { Size } from './ButtonGroupComponent';

interface IOverflowMenuWrapperProps {
  size: Size;
}

export const OverflowMenuWrapper = styled.div<IOverflowMenuWrapperProps>`
  ${({ size, theme }) => css`
    align-items: center;
    border-radius: ${theme.borderRadius};
    display: flex;
    justify-content: center;
    height: ${SIZE[size]}px;
    width: ${SIZE[size]}px;
  `};
`;

OverflowMenuWrapper.defaultProps = {
  theme: DEFAULT_THEME,
};
