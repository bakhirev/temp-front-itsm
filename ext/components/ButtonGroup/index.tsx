import React, { createRef, ElementType, FC, KeyboardEvent, MouseEvent } from 'react';

import { Badge } from '../Badge';
import { Button1, Button2 } from '../Typography';
import { OverflowMenu } from '../OverflowMenu';
import { KEY_CODES } from '../common';

import { AdditionalContentLayout } from './AdditionalContentLayout';
import { Button } from './Button';
import { ButtonGroupComponent } from './ButtonGroupComponent';
import { OverflowMenuWrapper } from './OverflowMenuWrapper';
import type { Kind } from './Button';
import type { Size } from './ButtonGroupComponent';

const ICON_SIZE = 24;
const ICON_SIZE_SMALL = 20;

export interface IButtonGroupProps {
  /** Идентификатор активной кнопки */
  activeButtonId: number | string;
  /** Имя класса для переопределения стилей */
  className?: string;
  /** data-test-id атрибут для тестирования компонента */
  dataTestId?: string;
  /** Внешний вид компонента */
  kind?: Kind;
  /** Список кнопок */
  list: Array<IButtonGroupItem>;
  /** Размер компонента */
  size?: Size;
  /** Коллбэк на переключение кнопки */
  onChange: (id: number | string) => void;
}

export interface IButtonGroupItem {
  badgeValue?: number;
  disabled?: boolean;
  hidden?: boolean;
  icon?: ElementType;
  id: number | string;
  label: string;
}

export const ButtonGroup: FC<IButtonGroupProps> = ({
  activeButtonId,
  className,
  dataTestId,
  kind = 'solid',
  list,
  size = 'medium',
  onChange,
}) => {
  const overflowMenuRef = createRef<any>();

  const handleMouseDown = (event: MouseEvent<HTMLDivElement>) => event.preventDefault();
  const handleOverflowMenuChange = ({ value }: any) => onChange(value);
  const handleOverflowMenuWrapperFocus = () => overflowMenuRef.current?.focus();

  const Label = size === 'medium' ? Button1 : Button2;

  const hiddenItems = list
    .filter(({ hidden }) => hidden)
    .map(({ disabled, id, label }) => ({
      isDisabled: disabled,
      isSelected: activeButtonId === id,
      label,
      value: id,
    }));

  const [activeButton] = list.filter(({ id }) => id === activeButtonId);

  return (
    <ButtonGroupComponent className={className} data-test-id={dataTestId} kind={kind} size={size}>
      {list.map(({ badgeValue, disabled = false, hidden = false, icon: Icon, id, label }) => {
        if (hidden) return null;

        const handleClick = () => {
          !disabled && onChange(id);
        };
        const handleKeyDown = (event: KeyboardEvent<HTMLDivElement>) => {
          if (event.keyCode === KEY_CODES.ENTER || event.keyCode === KEY_CODES.SPACEBAR) {
            event.preventDefault();
            !disabled && onChange(id);
          }
        };

        const getBadgeKind = () => {
          if (activeButtonId === id) {
            return 'neutral-light';
          }

          if (disabled) {
            return kind === 'solid' ? 'neutral-white-disabled' : 'neutral-light-disabled';
          }

          return kind === 'solid' ? 'neutral-white-inactive' : 'neutral-light-inactive';
        };

        const iconSize = size === 'medium' ? ICON_SIZE : ICON_SIZE_SMALL;
        return (
          <Button
            active={activeButtonId === id}
            disabled={disabled}
            key={id}
            kind={kind}
            size={size}
            tabIndex={disabled ? -1 : 1}
            onClick={handleClick}
            onKeyDown={handleKeyDown}
            onMouseDown={handleMouseDown}
          >
            {Icon && (
              <AdditionalContentLayout left>
                <Icon height={iconSize} width={iconSize} />
              </AdditionalContentLayout>
            )}
            <Label>{label}</Label>
            {badgeValue && (
              <AdditionalContentLayout right>
                <Badge
                  kind={getBadgeKind()}
                  size={size === 'medium' ? 'big' : 'small'}
                  value={badgeValue}
                />
              </AdditionalContentLayout>
            )}
          </Button>
        );
      })}

      {Boolean(hiddenItems.length) && (
        <OverflowMenuWrapper size={size} tabIndex={1} onFocus={handleOverflowMenuWrapperFocus}>
          <OverflowMenu
            innerRef={overflowMenuRef}
            list={hiddenItems}
            menu={{
              alignment: 'right',
              marginTop: 20,
            }}
            size={size === 'micro' ? 'medium' : 'big'}
            onChange={handleOverflowMenuChange}
            value={{
              value: activeButton.id,
              label: activeButton.label,
            }}
          />
        </OverflowMenuWrapper>
      )}
    </ButtonGroupComponent>
  );
};
