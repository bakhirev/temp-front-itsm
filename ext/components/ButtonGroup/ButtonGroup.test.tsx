import React from 'react';
import { shallow } from 'enzyme';

import { KEY_CODES } from '../common';
import { ButtonGroup } from '../ButtonGroup';

jest.mock('../common', () => ({
  DEFAULT_THEME: {
    borderRadius: '',
    color: {
      neutral: {
        50: '',
        30: '',
      },
    },
  },
  KEY_CODES: {
    ENTER: 13,
    SPACEBAR: 32,
  },
  Z_INDEXES: {
    TOOLTIP: 777,
  },
}));

describe('ButtonGroup', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const requiredProps = {
    activeButtonId: 0,
    list: [
      {
        id: 0,
        label: 'Foo',
      },
    ],
    onChange: jest.fn(),
  };

  it('should render component with required properties', () => {
    const wrapper = shallow(<ButtonGroup {...requiredProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render all items from list', () => {
    const item = requiredProps.list[0];
    const listItems = new Array(4).fill(item, 0, 4);

    const wrapper = shallow(<ButtonGroup {...requiredProps} list={listItems} />);
    expect(wrapper.children()).toHaveLength(listItems.length);
  });

  it('should call onChange if user clicks on button', () => {
    const wrapper = shallow(<ButtonGroup {...requiredProps} />);
    const item = wrapper.childAt(0);
    item.simulate('click');
    expect(requiredProps.onChange).toBeCalledTimes(1);
  });

  it("shouldn't call onChange if user clicks on disabled button", () => {
    const props = {
      ...requiredProps,
      list: [
        {
          ...requiredProps.list[0],
          disabled: true,
        },
      ],
    };
    const wrapper = shallow(<ButtonGroup {...props} />);
    const item = wrapper.childAt(0);
    item.simulate('click');
    expect(requiredProps.onChange).toBeCalledTimes(0);
  });

  it('should call onChange if user presses enter or space keys', () => {
    [KEY_CODES.ENTER, KEY_CODES.SPACEBAR].forEach((keyCode) => {
      requiredProps.onChange.mockClear();
      const wrapper = shallow(<ButtonGroup {...requiredProps} />);

      const item = wrapper.childAt(0);
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      item.simulate('keydown', { keyCode, preventDefault: () => {} });
      expect(requiredProps.onChange).toBeCalledTimes(1);
    });
  });

  it("shouldn't call onChange if user presses enter or space keys and button is disabled", () => {
    [KEY_CODES.ENTER, KEY_CODES.SPACEBAR].forEach((keyCode) => {
      const props = {
        ...requiredProps,
        list: [
          {
            ...requiredProps.list[0],
            disabled: true,
          },
        ],
      };

      const wrapper = shallow(<ButtonGroup {...props} />);

      const item = wrapper.childAt(0);
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      item.simulate('keydown', { keyCode, preventDefault: () => {} });
      expect(requiredProps.onChange).toBeCalledTimes(0);
    });
  });
});
