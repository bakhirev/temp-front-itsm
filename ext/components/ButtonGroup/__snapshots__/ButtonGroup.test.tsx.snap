// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup should render component with required properties 1`] = `
<styled.div
  kind="solid"
  size="medium"
  theme={
    Object {
      "borderRadius": "",
      "color": Object {
        "neutral": Object {
          "30": "",
          "50": "",
        },
      },
    }
  }
>
  <styled.div
    active={true}
    disabled={false}
    key="0"
    kind="solid"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    size="medium"
    tabIndex={1}
    theme={
      Object {
        "borderRadius": "",
        "color": Object {
          "neutral": Object {
            "30": "",
            "50": "",
          },
        },
      }
    }
  >
    <styled.div>
      Foo
    </styled.div>
  </styled.div>
</styled.div>
`;
